### PR TITLE
try to make the namespace controller faster

### PR DESCRIPTION
**Pull request by deads2k**
Trying to make the namespace controller faster. This
 1. runs the deletion of resources in parallel.  There's no reason to do this serially.
 1. Bumps the QPS for the deletion controller by 10x.  The default is 20 QPS and that means that deleting 100 individual items will take 5 seconds (best case).  Add discovery and list double checking and you get close to the 10+ seconds we're seeing.

@smarterclayton @derekwaynecarr

### CODE REVIEW

**Comment by smarterclayton** [comment]
Deletion controller uses bulk deletion, so in practice it is O(resourceTypes) rather than O(resources).

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go [hunk 1]
```diff
@@ -416,17 +421,42 @@ func (d *namespacedResourcesDeleter) deleteEachItem(
 		return nil
 	}
 	apiResource := metav1.APIResource{Name: gvr.Resource, Namespaced: true}
+
+	waitForAllDeletes := sync.WaitGroup{}
```
Comment (smarterclayton) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: Don't we have a util helper for this?

Comment (deads2k) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: > Don't we have a util helper for this?

For making a wait group or for spawning go funcs which write results to a local type in a channel?

File: pkg/controller/namespace/namespace_controller.go [hunk 2]
```diff
@@ -79,11 +71,7 @@ func NewNamespaceController(
 
 	// create the controller so we can inject the enqueue function
 	namespaceController := &NamespaceController{
-		kubeClient:                 kubeClient,
-		clientPool:                 clientPool,
```
Comment (smarterclayton) [inline pkg/controller/namespace/namespace_controller.go]: Were these just not refactored cleanly by previous output?

Comment (deads2k) [inline pkg/controller/namespace/namespace_controller.go]: > Were these just not refactored cleanly by previous output?

Looks like at some point the factorization of the controller changed and all the logic moved out into a different type which does the same thing.  I noticed it as I was checking the wiring

Comment (deads2k) [inline pkg/controller/namespace/namespace_controller.go]: I'd be in favor of eliminating the extra layer if its not used elsewhere; I didn't see other uses.  The current construction doesn't seem to be used for mocking.  I can't sort out why we have it.

Comment (smarterclayton) [inline pkg/controller/namespace/namespace_controller.go]: It's used in federation I think.

File: cmd/kube-controller-manager/app/core.go [hunk 3]
```diff
@@ -102,9 +103,16 @@ func startNamespaceController(ctx ControllerContext) (bool, error) {
...
+	nsKubeconfig := ctx.ClientBuilder.ConfigOrDie("namespace-controller")
+	nsKubeconfig.QPS *= 10
+	nsKubeconfig.Burst *= 10
+	namespaceKubeClient := clientset.NewForConfigOrDie(nsKubeconfig)
+	namespaceClientPool := dynamic.NewClientPool(nsKubeconfig, restMapper, dynamic.LegacyAPIPathResolverFunc)
+
 	// Find the list of namespaced resources via discovery that the namespace controller must manage
-	namespaceKubeClient := ctx.ClientBuilder.ClientOrDie("namespace-controller")
-	namespaceClientPool := dynamic.NewClientPool(ctx.ClientBuilder.ConfigOrDie("namespace-controller"), restMapper, dynamic.LegacyAPIPathResolverFunc)
 	// TODO: consider using a list-watch + cache here rather than polling
```
Comment (kargakis) [inline cmd/kube-controller-manager/app/core.go]: Is this still true?

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go [hunk 4]
```diff
@@ -440,7 +470,7 @@ func (d *namespacedResourcesDeleter) deleteAllContentForGroupVersionResource(
 	}
 	glog.V(5).Infof("namespace controller - deleteAllContentForGroupVersionResource - estimate - namespace: %s, gvr: %v, estimate: %v", namespace, gvr, estimate)
 
-	// get a client for this group version...
+	// get a client for this group version...  hread-safe
```
Comment (kargakis) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: typo

File: pkg/controller/namespace/deletion/namespaced_resources_deleter.go [hunk 5]
```diff
@@ -504,20 +533,52 @@ func (d *namespacedResourcesDeleter) deleteAllContent(
...
+	waitForAllResources := sync.WaitGroup{}
+	type result struct {
+		err      error
+		estimate int64
+	}
+	// make it big enough to handle all results without blocking
+	resultCh := make(chan result, len(groupVersionResources))
+	for key := range groupVersionResources {
+		gvr := key
+		// delete all the resources concurrently and the let the server sort them out
```
Comment (kargakis) [inline pkg/controller/namespace/deletion/namespaced_resources_deleter.go]: typo

**Comment by deads2k** [comment]
running in parallel just exacerbated qps problems.

**Pull request closed without merging**